File: 01-HashTables/unique_snowflakes.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 100000

typedef struct sf_node {
   int sf[6];
   struct sf_node *next;
} sf_node;

int code(int sf[]) {
   int sum = 0;
   for (int i=0; i<6; i++)
      sum += sf[i];
   return sum % SIZE;
}

int i_right(int sf1[], int sf2[], int start) {
   for (int i=0; i<6; i++) {
      if (sf1[i] != sf2[(start+i) % 6])
         return 0;
   }
   return 1;
}

int i_left(int sf1[], int sf2[], int start) {
   int sf2_i = 0;
   for (int i=0; i<6; i++) {
      sf2_i = start - i;
      if (sf2_i < 0) sf2_i += 6;
      if (sf1[i] != sf2[sf2_i])
         return 0;
   }
   return 1;
}

int are_ident(int sf1[], int sf2[]) {
   for (int i=0; i<6; i++) {
      if (i_right(sf1, sf2, i))
         return 1;
      if (i_left(sf1, sf2, i))
         return 1;
   }
   return 0;
}

void find_ident(sf_node *sfs[]) {
   sf_node *node1, *node2;
   for (int i=0; i<SIZE; i++) {
      node1 = sfs[i];
      while (node1 != NULL) {
         node2 = node1->next;
         while (node2 != NULL) {
            if (are_ident(node1->sf, node2->sf)) {
               printf("Twin snowflakes found.\n");
               return;
            }
            node2 = node2->next;
         }
         node1 = node1->next;
      }
   }
   printf("No two snowflakes are alike.\n");
}

int main(void) {
   static sf_node *sfs[SIZE] = {NULL};
   sf_node *sf;
   int n, sf_code = 0;
   scanf("%d", &n);
   for (int i=0; i<n; i++) {
      sf = malloc(sizeof(sf_node));
      for (int j=0; j<6; j++)
         scanf("%d", &sf->sf[j]);
      sf_code = code(sf->sf);
      sf->next = sfs[sf_code];
      sfs[sf_code] = sf;
   }
   find_ident(sfs);
   return 0;
}

File: 01-HashTables/unique_snowflakes.ml
let sf_hash = Hashtbl.create 100000;;
Hashtbl.add sf_hash [-1] (Array.make 6 0)
let found = ref false

let code sf =
  (Array.fold_right (fun x y -> x + y) sf 0) mod 100000

let shiftR (sf1: int array) sf2 j =
  let equals = ref true in
  for i = 0 to 5 do
    equals := sf1.(i) = sf2.((i+j) mod 6);
  done;
  !equals

let shiftL (sf1: int array) sf2 j =
  let equals = ref true in
  for i = 0 to 5 do
    equals := sf1.(i) = sf2.((j-i+6) mod 6);
  done;
  !equals

let rec ident (sf1: int array) (sf2: int array) indices = match indices with
  | [] -> false
  | j::js -> shiftR sf1 sf2 j || shiftL sf1 sf2 j || ident sf1 sf2 js 

let are_ident sf1 sf2 =
  let indices = ref [] in
    for i = 0 to 5 do
      if sf1.(0) = sf2.(i)
        then indices := i::!indices
    done;
    ident sf1 sf2 !indices

let rec check_matches matches sf = match matches with
  | [] -> false
  | m::ms -> are_ident m sf || check_matches ms sf

let find_ident hash sf =
  let matches = Hashtbl.find_all sf_hash [hash] in
  check_matches matches sf

let main () =
  let n = read_int () in
  for i = 1 to n do
    let sf = Array.make 6 0 in
    let line = read_line () in
    Scanf.sscanf line "%d %d %d %d %d %d" (fun a b c d e f -> 
      sf.(0) <- a; sf.(1) <- b; sf.(2) <- c;
      sf.(3) <- d; sf.(4) <- e; sf.(5) <- f);
    let hash = code sf in
    if Hashtbl.mem sf_hash [hash]
      then found := find_ident hash sf;
    Hashtbl.add sf_hash [hash] sf;
    if !found then print_endline "Twin snowflakes found.";
    if !found then exit 0
  done;
  print_endline "No two snowflakes are alike.";;

main ()